### PR TITLE
Add tests of two contexts

### PR DESCRIPTION
**Pull request by harry-shepherd**
I have added an example to xios_examples demonstrating the ability to use XIOS contexts to stop output part way through the model run.

**Brief description**
There are two identical contexts defined in the `main_stop_*.xml` files, stopping after 5 and 10 timesteps respectively. The only difference is the name of the output files differ.

The program is controlled from the namelist file `context.nml` where we can have up to 4 different stop points in this case (although in the example there are currently two). They need a name, and a timestep on which we want to load the context.
These are then read into a list of "context_struct" types, (defined in `custom_type_mod.F90`) in `context_def_mod.F90`
 
The actual test code is in `context_def_test.F90`, where we call the `load_contexts` subroutine from `context_def_mod`. We then initialize the contexts in a loop, to ensure that they are self consistent, and place their handles into the context_struct type, along with setting a flag `is_avaliable` to `.true.`, to show that these can be used.
 
 
When we timestep, again we loop over the contexts, check that it is available to us, send fields as we normally would. I haven't, but it may be worth using the xios functionality to check the field is available on this context and this timestep to make sure we aren't trying to send too much data around. If the context has reached its end point, we close it, and set the is_available flag to `.false.` to make sure we dont attempt to use it again.

### CODE REVIEW

File: xios_examples/context_test/context_def_mod.F90 [hunk 1]
```diff
@@ -0,0 +1,46 @@
+module context_def_mod
+
+!use xios
```
Comment (mo-marqh) [inline xios_examples/context_test/context_def_mod.F90:3]: this line appears commented out, which is a bit confusing.
can it be removed?

File: xios_examples/context_test/README [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+The program is controlled from the namelist file context.nml where we can have up to 4 different stop points in this case (although in the example there are currently two). They need a name, and a timestep on which we want to load the context.
+These are then read into a list of "context_struct" types, (defined in custom_type_mod.F90) in context_def_mod.F90
+
+The actual test code is in context_def_test.F90, where we call the load_contexts subroutine from context_def_mod. We then initialize the contexts in a loop, to ensure that they are self consistent, and place their handles into the context_struct type, along with setting a flag is_avaliable to .true., to show that these can be used.
+
+When we timestep, again we loop over the contexts, check that it is available to us, send fields as we normally would. I haven't, but it may be worth using the xios functionality to check the field is available on this context and this timestep to make sure we aren't trying to send too much data around. If the context has reached its end point, we close it, and set the is_available flag to .false. to make sure we dont attempt to use it again.
```
Comment (mo-marqh) [inline xios_examples/context_test/README:13]: empty line at end of file is preferred please (else github gives warnings)

File: xios_examples/context_test/context_def_mod.F90 [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+  character(len=20) :: context_names(4) = ''
+  integer :: closing_timesteps(4) = 0
+
+  integer :: file_unit, io_stat
+  integer :: i
+
+  type(context_struct) :: my_context_struct
+
+  namelist /CUSTOM_CONTEXTS/ n_contexts, context_names, closing_timesteps
```
Comment (mo-marqh) [inline xios_examples/context_test/context_def_mod.F90:28]: some code comments here would be really helpful please

File: xios_examples/context_test/custom_type_mod.F90 [hunk 4]
```diff
@@ -0,0 +1,15 @@
+module custom_type_mod
+
```
Comment (mo-marqh) [inline xios_examples/context_test/custom_type_mod.F90:2]: a code comment explaining what this module is here for would be really helpful please

File: xios_examples/context_test/context_def_mod.F90 [hunk 5]
```diff
@@ -0,0 +1,46 @@
+module context_def_mod
+
```
Comment (mo-marqh) [inline xios_examples/context_test/context_def_mod.F90:2]: a code comment explaining what this module is here for would be really helpful please

**Review by mo-marqh — COMMENTED** [review]
this looks good @harry-shepherd 

there are a few places where a bit more commenting would help me, as i haven't quite understood the reasons for the different Fortran modules. i don't think much detail is required here, just a clue or two for the future reader

tests are running and passing on all XIOS versions

**Review by mo-marqh — APPROVED** [review]
looks great, thank you

**Comment by mo-marqh** [comment]
@harry-shepherd 
this is ready to merge (which i'll do)

i just wanted to check how valuable the separate commits are here.

it looks like it's valid to `Squash & Merge` into a single coherent change to me, are you content with this?

**Comment by harry-shepherd** [comment]
> @harry-shepherd this is ready to merge (which i'll do)
> 
> i just wanted to check how valuable the separate commits are here.
> 
> it looks like it's valid to `Squash & Merge` into a single coherent change to me, are you content with this?
I am happy for you to `Squash & Merge`, Thanks

**Pull request merged**